cmd/allocator/main.go

```diff
@@ -245,16 +246,14 @@ func main() {
 	signals.NewSigTermHandler(func() {
 		logger.Info("Pod shutdown has been requested, failing readiness check")
 		podReady = false
+		grpcHealth.Shutdown()
 		time.Sleep(conf.ReadinessShutdownDuration)
-		cancelCtx()
-		logger.Infof("Readiness shutdown duration has passed, context cancelled")
-		time.Sleep(1 * time.Second) // allow a brief time for cleanup, but force exit if main doesn't
-		os.Exit(0)
+		cancelListenCtx()
 	})
 
 	grpcUnallocatedStatusCode := grpcCodeFromHTTPStatus(conf.httpUnallocatedStatusCode)
 
-	h := newServiceHandler(ctx, kubeClient, agonesClient, health, conf.MTLSDisabled, conf.TLSDisabled, conf.remoteAllocationTimeout, conf.totalRemoteAllocationTimeout, conf.allocationBatchWaitTime, grpcUnallocatedStatusCode)
+	h := newServiceHandler(context.Background(), kubeClient, agonesClient, health, conf.MTLSDisabled, conf.TLSDisabled, conf.remoteAllocationTimeout, conf.totalRemoteAllocationTimeout, conf.allocationBatchWaitTime, grpcUnallocatedStatusCode)
```

Comment:
passing in `ctx.Background()` seems not ideal here - it's passed through to all the worker queues and informers etc - so none of them will shutdown on termination.

Any reason we can't use `listenCtx` ?

Reply:
Part of the point was to separate the contexts between the context for all the listeners (so they could shutdown) and all the "forward" contexts (workers, informers, etc.). The idea is to drain the listeners, then the forward context should just have nothing to do shortly.

Reply:
FWIW the right / more common way to handle this is just to have a channel instead of a separate context, but that doesn't fit the runner interface that everything uses..

Reply:
I went ahead and separated the contexts and added a TODO to make this more obvious.